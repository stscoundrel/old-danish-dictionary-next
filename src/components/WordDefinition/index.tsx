import { capitalize } from 'lib/utils/strings'
import { lettersToRunes } from 'futhork'
import { DictionaryEntry } from 'lib/models/dictionary'
import { addAbbreviationsToContent } from 'lib/services/abbreviations/client'
import Abbreviations from 'components/Abbreviations'
import SimilarEntries from 'components/SimilarEntries'
import { Abbreviation } from 'lib/services/abbreviations/model'
import styles from './WordDefinition.module.scss'

interface WordDefinitionProps{
  entry: DictionaryEntry,
  similarEntries: DictionaryEntry[],
  abbreviations: Abbreviation[],
}

export default function WordDefinition({
  entry, similarEntries, abbreviations,
}: WordDefinitionProps) {
  const {
    headword, definitions,
  } = entry

  return (
    <article className={styles.section}>
      <header>
        <h1 lang="non">{capitalize(headword)}</h1>

        <small className={styles.subHeading}>
          Old Danish Dictionary - {headword.toLowerCase()}
        </small>
        <p>Meaning of Old Danish word <em>&quot;{headword}&quot;</em>,
        as defined by Otto Kalkar&apos;s Dictionary of Old Danish language.</p>
      </header>

      <p>The Old Danish word <strong><dfn className="capitalize">{headword}</dfn></strong> can mean:</p>
      <ul className={styles.definitionList}>
      {definitions.map((definition, index) => (
          <li
            key={`definition-${index}`}
            className={styles.definitionItem}
            lang="swe"
            dangerouslySetInnerHTML={{
              __html: addAbbreviationsToContent(definition, abbreviations),
            } }
          />
      ))}
      </ul>

      <p>Possible runic inscription in <em>Medieval Futhork</em>:
        <span className={styles.rune}>{ lettersToRunes(headword) }</span><br />
      <small>Medieval Runes were used in Denmark from 12th to 17th centuries.</small>
      </p>

      <SimilarEntries entries={similarEntries} />
      <Abbreviations abbreviations={abbreviations} />
    </article>
  )
}
