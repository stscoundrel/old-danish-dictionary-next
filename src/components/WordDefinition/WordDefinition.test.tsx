import { DictionaryEntry } from 'lib/models/dictionary'
import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import { Crosslink, DictionarySource } from 'scandinavian-dictionary-crosslinker'
import WordDefinition from './index'

const entry: DictionaryEntry = {
  headword: 'Abe',
  definitions: [
    'no.',
    "1) dyrenavn, hunk. (cv sv. apa.); som aben tit skeer, naar hun vil gøre effter det, hun seer. Herm. ' Weigere. 215, 228%; lige som man ick kunde aben kende, naar hun sig før i Aarons klæder. Ranch. 29; løb løss en tam abe, der mand lockede ad hende. Hvitf VIL 1; aben, som bilder sig ind, at hendis børn er altijd de smuckeste. S. Terkelsen, Astreæ Sjungek. IL. fort. —",
    '2) fjfr. isl. api.) dåre; wthen han seer ther tijl ful- breth, tha holles han forenabe., Rimkr. p%; giør du Joab, Herre, til din abe, at hans anslag ey due. Ranch. 95; en nar, den, som holder for- meget af sine børn. Moth.',
  ],
  slug: 'abe',
}

const abbreviations = [
  { abbreviation: 'no.', explanation: 'navneord (substantivum).' },
  { abbreviation: 'hunk.', explanation: 'hunkøn (femininum).' },
  { abbreviation: 'sv.', explanation: 'svensk.' },
  { abbreviation: 't.', explanation: 'tysk.' },
  { abbreviation: 'isl.', explanation: 'islandsk.' },
  { abbreviation: 'n.', explanation: 'norsk.' },
]

const crosslinks: Crosslink[] = [
  {
    url: 'https://old-swedish-dictionary.vercel.app/word/dag',
    source: DictionarySource.OldSwedish,
  },
]

describe('WordDefinition component', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(
      <WordDefinition
        entry={entry}
        abbreviations={abbreviations}
        similarEntries={[entry]}
        crosslinks={crosslinks}
      />,
    )
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <WordDefinition
        entry={entry}
        similarEntries={[entry]}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct title', () => {
    const tree = renderer.create(
      <WordDefinition
        entry={entry}
        similarEntries={[entry]}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    )
    const { root } = tree

    expect(root.findByType('h1').children).toEqual(['Abe'])
  })
})
