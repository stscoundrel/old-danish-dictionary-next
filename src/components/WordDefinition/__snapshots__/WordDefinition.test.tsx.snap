// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordDefinition component Matches snapshot 1`] = `
<article
  className="section"
>
  <header>
    <h1
      lang="non"
    >
      Abe
    </h1>
    <small
      className="subHeading"
    >
      Old Danish Dictionary - 
      abe
    </small>
    <p>
      Meaning of Old Danish word 
      <em>
        "
        Abe
        "
      </em>
      , as defined by Otto Kalkar's Dictionary of Old Danish language.
    </p>
  </header>
  <p>
    The Old Danish word 
    <strong>
      <dfn
        className="capitalize"
      >
        Abe
      </dfn>
    </strong>
     can mean:
  </p>
  <ul
    className="definitionList"
  >
    <li
      className="definitionItem"
      dangerouslySetInnerHTML={
        {
          "__html": "<abbr title="navneord (substantivum).">no.</abbr>",
        }
      }
      lang="swe"
    />
    <li
      className="definitionItem"
      dangerouslySetInnerHTML={
        {
          "__html": "1) dyrenavn, <abbr title="hunkøn (femininum).">hunk.</abbr> (cv <abbr title="svensk.">sv.</abbr> apa.); som aben tit skeer, naar hun vil gøre effter det, hun seer. Herm. ' Weigere. 215, 228%; lige som man ick kunde aben kende, naar hun sig før i Aarons klæder. Ranch. 29; løb løss en tam abe, der mand lockede ad hende. Hvitf VIL 1; aben, som bilder sig ind, at hendis børn er altijd de smuckeste. S. Terkelsen, Astreæ Sjungek. IL. for<abbr title="tysk.">t.</abbr> —",
        }
      }
      lang="swe"
    />
    <li
      className="definitionItem"
      dangerouslySetInnerHTML={
        {
          "__html": "2) fjfr. <abbr title="islandsk.">isl.</abbr> api.) dåre; wthen han seer ther tijl ful- breth, tha holles han forenabe., Rimkr. p%; giør du Joab, Herre, til din abe, at hans anslag ey due. Ranch. 95; en nar, den, som holder for- meget af sine bør<abbr title="norsk.">n.</abbr> Moth.",
        }
      }
      lang="swe"
    />
  </ul>
  <p>
    Possible runic inscription in 
    <em>
      Medieval Futhork
    </em>
    :
    <span
      className="rune"
    >
      ᛆᛒᚽ
    </span>
    <br />
    <small>
      Medieval Runes were used in Denmark from 12th to 17th centuries.
    </small>
  </p>
  <h4>
    Similar entries:
  </h4>
  <ul>
    <li>
      <a
        className="link"
        href="/word/abe"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Abe (I)
      </a>
    </li>
  </ul>
  <div
    className="abbreviations"
  >
    <div
      className="column"
    >
      <h4>
        Abbreviations used:
      </h4>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            no.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          navneord (substantivum).
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            hunk.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          hunkøn (femininum).
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            sv.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          svensk.
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            t.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          tysk.
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            isl.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          islandsk.
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            n.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          norsk.
        </dd>
      </dl>
    </div>
  </div>
  <br />
  <h4>
    Also available in related dictionaries:
  </h4>
  <p>
    This headword also appears in dictionaries of other languages closely related to Old Danish.
  </p>
  <ul
    className="list"
  >
    <li
      className="listItem"
    >
      <a
        href="https://old-swedish-dictionary.vercel.app/word/dag"
        rel="noopener noreferrer"
        target="_blank"
      >
        Old Swedish - K.F Söderwall's Dictionary
      </a>
    </li>
  </ul>
</article>
`;
